spec.html

```diff
@@ -6419,135 +6419,32 @@ <h1>GetGlobalObject ( )</h1>
-        1. Let _callerContext_ be the running execution context.
-        1. Let _callerRealm_ be _callerContext_'s Realm.
-        1. Let _callerScriptOrModule_ be _callerContext_'s ScriptOrModule.
-        1. Let _pending_ be PendingJob { [[Job]]: _job_, [[Arguments]]: _arguments_, [[Realm]]: _callerRealm_, [[ScriptOrModule]]: _callerScriptOrModule_, [[HostDefined]]: *undefined* }.
-        1. Perform any implementation or host environment defined processing of _pending_. This may include modifying the [[HostDefined]] field or any other field of _pending_.
-        1. Add _pending_ at the back of the Job Queue named by _queueName_.
-        1. Return NormalCompletion(~empty~).
-      </emu-alg>
+    <p>Jobs are scheduled for execution by ECMAScript host environments with abstract operations such as HostEnqueuePromiseJob. Such operations accept two parameters: _job_, an abstract operation, and _arguments_, a list of ECMAScript values. These abstract operations schedule _job_ to be performed with _arguments_ at some future time. Their implementations must conform to the following requirements:</p>
+
+    <ul>
+      <li>At some future point in time, when there is no running execution context and the execution context stack is empty, the implementation must:
+        <ol>
+          <li>Push an execution context onto the execution context stack.</li>
+          <li>Perform any implementation-defined preparation steps.</li>
+          <li>Perform the abstract operation given by _job_, passing the arguments _arguments_.</li>
+          <li>Perform any implementation-defined cleanup steps.</li>
+          <li>Pop the previously-pushed execution context from the execution context stack.</li>
+        </ol>
+      </li>
+      <li>Only one Job may be actively undergoing evaluation at any point in time.</li>
+      <li>Once evaluation of a Job starts, it must run to completion before evaluation of any other Job starts.</li>
+      <li>The evaluation of any Jobs enqueueed via this or other host job scheduling algorithms must complete before evaluation of the Job starts.</li>
+    </ul>
+
+    <emu-clause id="sec-hostenqueuepromisejob" aoid="HostEnqueuePromiseJob">
+      <h1>HostEnqueuePromiseJob ( _job_, _arguments_ )</h1>
+
+      <p>HostEnqueuePromiseJob is a host-defined abstract operation that schedules the abstract operation indicated by _job_ to be performed, with the arguments _arguments_, at some future time. The Jobs used with this algorithm are intended to be related to the handling of Promises, or otherwise, to be scheduled with equal priority to Promise handling operations.</p>
+
+      <p>The implementation of HostEnqueuePromiseJob must conform to the requirements in <emu-xref href="#sec-jobs-and-job-queues"></emu-xref>. Additionally, Jobs must be scheduled in FIFO order, with Jobs running in the same order as the HostEnqueuePromiseJob invocations which scheduled them.</p>
```

Comment:
"Jobs must be scheduled in FIFO order" seems like a general requirement that belongs in the previous section (and I think the previous section already contains it, in different words?).

Reply:
The motivation for putting it in this section was that not all host job hooks will use it. For example, I don't think we want this to apply to WeakRef finalization callbacks. However, we will want it to apply to Atomics.waitAsync.

Reply:
The above general requirements state

> The evaluation of any Jobs enqueueed via this or other host job scheduling algorithms must complete before evaluation of the Job starts.

This seems like the same requirement, applied to all jobs?

Reply:
I think the above requirement is more about run-to-completion (which applies to all host job hooks), and this requirement is about FIFO ordering (which only applies to some of them).

Reply:
I think I can see that reading now. Any jobs that have not yet started, don't trigger that clause.

In that case I think that clause should be deleted, since it comes right before an IMO-clearer statement:

> Once evaluation of a Job starts, it must run to completion before evaluation of any other Job starts.

Reply:
OK, deleted; I agree that it's a bit redundant.

Reply:
This still hasn’t been deleted; is a new diff expected?

Reply:
What used to be at 6439 was deleted, following @domenic's recommendation.

Reply:
Does the phrase “in FIFO order” need to appear? It seems redundant with the prose immediately following it.

Reply:
It's slightly redundant for clarity. FIFO was present in the previous text, and the FIFO ordering was also clear from the RunJobs algorithm, so the wording here is analogous to that. If it's a big deal for you, I could move part of it to a note, but I think it's OK to have a single sentence with words that agree with each other.